pydrofoil/real.py

```diff
@@ -0,0 +1,115 @@
+import sys
+from rpython.rlib.rbigint import rbigint, _divrem as bigint_divrem
+from rpython.rlib.rarithmetic import r_uint, intmask, string_to_int, ovfcheck, \
+        int_c_div, int_c_mod, r_ulonglong
+
+MAXINT = 2**63-1
+MININT = -2**63
+
+class Real(object):
+    def __init__(self, num, den):
+        gcd_val = num.gcd(den)
+        self.num = num.div(gcd_val)
+        self.den = den.div(gcd_val)
+
+    @staticmethod
+    def fromint(num, den=1):
+        assert not den==0, "denominator cannot be 0"
+        if num == MININT or den == MININT:
+            num = rbigint.fromint(num)
+            den = rbigint.fromint(den)
+            sign = num.sign * den.sign
+            num = num.abs().int_mul(sign)
+            den = den.abs()
+        elif num < MININT or num > MAXINT or den < MININT or den > MAXINT:
+            assert False, "input is out of range of INT"
+        else:
+            sign = num/abs(num)*den/abs(den)
+            num = rbigint.fromint(abs(num)*sign)
+            den = rbigint.fromint(abs(den))
+        return Real(num, den)
+    
+    @staticmethod
+    def fromstr(str):
+        from rpython.rlib.rstring import strip_spaces
+        s = strip_spaces(str)
+        decimalpos = s.find(".")
+        num = rbigint.fromstr(s[:decimalpos]+s[decimalpos+1:], 10) if decimalpos < len(s)-1 else rbigint.fromstr(s, 10)
```

Comment:
I think this produces incorrect results if there is no decimal point at all. what does `Real.fromstr("1")` return? I am fine with simply adding an assert that there must be a `.` though, like this: `assert decimalpos >= 0` (`str.find` returns a negative number if the search string isn't found).